cpp/celeborn/conf/BaseConf.h

```diff
@@ -0,0 +1,263 @@
+    auto val = get(key);
+    if (val.hasValue()) {
+      return folly::to<T>(val.value());
+    } else {
+      return folly::none;
+    }
+  }
+
+  template <typename T>
+  T get(const std::string& key, const T& defaultValue) const {
+    auto val = get(key);
+    if (val.hasValue()) {
+      return folly::to<T>(val.value());
+    } else {
+      return defaultValue;
+    }
+  }
+
+  virtual bool isValueExists(const std::string& key) const = 0;
+
+  virtual const std::unordered_map<std::string, std::string>& values() const {
+    CELEBORN_UNSUPPORTED("method values() is not supported by this config");
+  }
+
+  virtual std::unordered_map<std::string, std::string> valuesCopy() const {
+    CELEBORN_UNSUPPORTED("method valuesCopy() is not supported by this config");
+  }
+};
+
+namespace core {
+
+class MemConfig : public Config {
+ public:
+  explicit MemConfig(const std::unordered_map<std::string, std::string>& values)
+      : values_(values) {}
+
+  explicit MemConfig() : values_{} {}
+
+  explicit MemConfig(std::unordered_map<std::string, std::string>&& values)
+      : values_(std::move(values)) {}
+
+  folly::Optional<std::string> get(const std::string& key) const override;
+
+  bool isValueExists(const std::string& key) const override;
+
+  const std::unordered_map<std::string, std::string>& values() const override {
+    return values_;
+  }
+
+  std::unordered_map<std::string, std::string> valuesCopy() const override {
+    return values_;
+  }
+
+ private:
+  std::unordered_map<std::string, std::string> values_;
+};
+
+/// In-memory config allowing changing properties at runtime.
+class MemConfigMutable : public Config {
+ public:
+  explicit MemConfigMutable(
+      const std::unordered_map<std::string, std::string>& values)
+      : values_(values) {}
+
+  explicit MemConfigMutable() : values_{} {}
+
+  explicit MemConfigMutable(
+      std::unordered_map<std::string, std::string>&& values)
+      : values_(std::move(values)) {}
+
+  folly::Optional<std::string> get(const std::string& key) const override;
+
+  bool isValueExists(const std::string& key) const override;
+
+  const std::unordered_map<std::string, std::string>& values() const override {
+    CELEBORN_UNSUPPORTED(
+        "Mutable config cannot return unprotected reference to values.");
+    return *values_.rlock();
+  }
+
+  std::unordered_map<std::string, std::string> valuesCopy() const override {
+    return *values_.rlock();
+  }
+
+  /// Adds or replaces value at the given key. Can be used by debugging or
+  /// testing code.
+  void setValue(const std::string& key, const std::string& value) {
+    (*values_.wlock())[key] = value;
+  }
+
+ private:
+  folly::Synchronized<std::unordered_map<std::string, std::string>> values_;
+};
+
+} // namespace core
+
+class BaseConf {
+ public:
+  // Setting this to 'true' makes configs modifiable via server operations.
+  static constexpr std::string_view kMutableConfig{"mutable-config"};
```

Comment:
Why should we need immutable in-memory configs?

Reply:
For a general-purpose config system, immutable configs are required because many system-wide configs are not modifiable, such as service thread num, execution mode, etc. For these configs, immutability is by design, and modifying them at runtime would not take effect. The Mutability design here is to enforce the configs to be not modifiable.
If this is not aligned to Java end's behavior, we could modify the default mode to mutable. Let me know if the modification is required, thanks a lot.

Reply:
In the Java end, we have mutable configs by default. I think just changing the default configs to mutable should be enough.